### PR TITLE
refactor llamahub download

### PR DESCRIPTION
**Pull request by jerryjliu**
refactor llama-hub downloads into a set of shared modules

now we can load both readers and tools very easily (as well as any future modules)

### CODE REVIEW

File: docs/examples/llama_hub/llama_hub.ipynb [hunk 1]
```diff
@@ -0,0 +1,227 @@
+{
```
Comment (logan-markewich) [inline docs/examples/llama_hub/llama_hub.ipynb:1]: <p>nit: Demonstration</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/run-llama/llama_index/pull/8669/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='0'/>

**Review by logan-markewich — APPROVED** [review]
This looks good to me (just some linting changes). Although at some point we really need to pick reader vs loader 😅

**Comment by jerryjliu** [comment]
> This looks good to me (just some linting changes). Although at some point we really need to pick reader vs loader 😅

good point...we can also alias `download_loader` with `download_reader

**Comment by mclassen** [comment]
I'm getting an error on windows related to the download modulel:

Invalid URL `'https:\\raw.githubusercontent.com\\run-llama\\llama-hub\\main\\llama_hub/gpt_repo/requirements.txt'`: No host supplied

It seems that in `download_utils.py` `download_llama_module()` `llama_hub_url `is still ok, e.g. it is `https://raw.githubusercontent.com/run-llama/llama-hub/main/llama_hub`

But in later calls, somehow the first part will get converted as if it was a string that is "somehow converted into a windows path and escaped", I can't really make sense of it...

**Pull request merged**